src/coreComponents/constitutive/fluid/PVTFunctions/CO2SolubilityFunction.cpp

```diff
@@ -16,379 +16,258 @@
-
-  m = salinity;
+  { 28.9447706, -0.0354581768, -4770.67077, 1.02782768e-5, 33.8126098, 9.04037140e-3,
+    -1.14934031e-3, -0.307405726, -0.0907301486, 9.32713393e-4, 0 };
+  constexpr real64 lambda[] = { -0.411370585, 6.07632013e-4, 97.5347708, 0, 0, 0, 0, -0.0237622469, 0.0170656236, 0, 1.41335834e-5 };
+  constexpr real64 zeta[] = { 3.36389723e-4, -1.98298980e-5, 0, 0, 0, 0, 0, 2.12220830e-3, -5.24873303e-3, 0, 0 };
 
-  for( localIndex i = 0; i < pressure.size(); ++i )
+  for( localIndex i = 0; i < tableCoords.nPressures(); ++i )
   {
+    real64 const P = tableCoords.getPressure( i ) / P_Pa_f;
 
-    P = pressure[i] / P_Pa_f;
-
-    for( localIndex j = 0; j < temperature.size(); ++j )
+    for( localIndex j = 0; j < tableCoords.nTemperatures(); ++j )
     {
+      real64 const T = tableCoords.getTemperature( j );
 
-      T = temperature[j];
-
-      CO2Solubility( T, P, V_r, &ff );
-
-      logK = Par( T+T_K_f, P, mu ) - logF( T, P, V_r ) + 2*Par( T+T_K_f, P, lambda )*m + Par( T+T_K_f, P, zeta )*m*m;
+      // compute reduced volume by solving the CO2 equation of state
+      real64 V_r = 0.0;
+      CO2SolubilityFunction( T, P, V_r, &detail::ff );
 
-      y_CO2 = (P - PWater( T ))/P;
-
-      solubiltiy[i][j] = y_CO2 * P / exp( logK );
+      // compute equation (6) of Duan and Sun (2003)
+      real64 const logK = detail::Par( T+T_K_f, P, mu )
+                          - detail::logF( T, P, V_r )
+                          + 2*detail::Par( T+T_K_f, P, lambda ) * salinity
+                          + detail::Par( T+T_K_f, P, zeta ) * salinity * salinity;
 
+      // mole fraction of CO2 in vapor phase, equation (4) of Duan and Sun (2003)
+      real64 const y_CO2 = (P - detail::PWater( T ))/P;
+      values[j*tableCoords.nPressures()+i] = y_CO2 * P / exp( logK );
     }
-
   }
-
 }
 
-
-CO2SolubilityFunction::CO2SolubilityFunction( string_array const & inputPara,
-                                              string_array const & phaseNames,
-                                              string_array const & componentNames,
-                                              real64_array const & componentMolarWeight ):
-  FlashModel( inputPara[1], componentNames, componentMolarWeight )
+void CO2Solubility::CO2SolubilityFunction( real64 const & T,
+                                           real64 const & P,
+                                           real64 & V_r,
+                                           real64 (*f)( real64 const & x1, real64 const & x2, real64 const & x3 ) )
 {
 
-  bool notFound = 1;
-
-  for( localIndex i = 0; i < componentNames.size(); ++i )
-  {
-
-    if( componentNames[i] == "CO2" || componentNames[i] == "co2" )
-    {
-      m_CO2Index = i;
-      notFound = 0;
-      break;
-    }
-
-  }
-
-  GEOSX_ERROR_IF( notFound, "Component CO2 is not found!" );
+  constexpr real64 eps = 1e-9;
+  constexpr real64 dx = 1e-10;
+  int count = 0;
+  real64 Vr_int = 0.05;
```

Comment:
Is there any way to let the user define those parameters (you can keep these defaults if you want obviously).

Reply:
Here I would say that it is not a good idea to let the user choose this type of parameters. I would prefer to let the (expert) users take GEOSX and recompile with a better parameter if needed.

Reply:
I think you're going wrong way here.

If I need to test another parameter set I need to spend hours connecting to the remote HPC, set the build env, build and deploy, just to test?
And then I have to re-deploy with all the Quality Assurance process, the tickets with the outsourced team...
And after that, what is the reference version ? Not the same code base, not the same hash...
Imagine that two users have incompatible parameters, do you have to deal with two different versions?

Reply:
Done, now the tolerance can be chosen from the input file.